libdrgn/iterators.c

```diff
@@ -0,0 +1,461 @@
+	}
+	frames_deinit(&iter->frames);
+}
+
+struct drgn_error *idr_iter_init(const struct drgn_object *idr,
+				 struct idr_iter *ret)
+{
+	struct drgn_object idr_rt;
+	struct drgn_object idr_base;
+	struct drgn_error *err;
+	drgn_object_init(&idr_rt, drgn_object_program(idr));
+	drgn_object_init(&idr_base, drgn_object_program(idr));
+
+	err = drgn_object_member(&idr_base, idr, "idr_base");
+	if (!err) {
+		err = drgn_object_read_unsigned(&idr_base, &ret->base);
+		if (err)
+			goto out;
+	} else if (err->code == DRGN_ERROR_LOOKUP) {
+		drgn_error_destroy(err);
+		err = NULL;
+		ret->base = 0;
+	} else {
+		goto out;
+	}
+
+	err = drgn_object_member(&idr_rt, idr, "idr_rt");
+	if (err)
+		goto out;
+
+	err = radix_tree_iter_init(drgn_object_program(idr), &idr_rt,
+				   &ret->iter);
+out:
+	drgn_object_deinit(&idr_rt);
+	drgn_object_deinit(&idr_base);
+	return err;
+}
+
+struct drgn_error *idr_iter_next(struct idr_iter *iter, idr_iter_entry_t **ret)
+{
+	struct drgn_error *err = radix_tree_iter_next(&iter->iter, ret);
+	if (!err && *ret)
+		(*ret)->index += iter->base;
+	return err;
+}
+
+void idr_iter_deinit(struct idr_iter *iter)
+{
+	radix_tree_iter_deinit(&iter->iter);
+}
+
+static struct drgn_error *pid_iter_init_old_kernel(struct drgn_program *prog,
+						   const struct drgn_object *ns,
+						   struct pid_iter *ret)
+{
+	struct drgn_error *err;
+	ret->has_idr = false;
+	ret->index = -1;
+	drgn_object_init(&ret->pid_hash, prog);
+	drgn_object_init(&ret->pos, prog);
+	drgn_object_init(&ret->ns, prog);
+	drgn_object_init(&ret->entry, prog);
+	err = drgn_program_find_object(prog, "pid_hash", NULL,
+				       DRGN_FIND_OBJECT_ANY, &ret->pid_hash);
+	if (err)
+		return err;
+	struct drgn_qualified_type void_star_type;
+	err = drgn_program_find_type(prog, "void *", NULL, &void_star_type);
+	if (err)
+		return err;
+	err = drgn_object_set_unsigned(&ret->pos, void_star_type, 0, 0);
+	if (err)
+		return err;
+	err = drgn_object_copy(&ret->ns, ns);
+	if (err)
+		return err;
+	err = drgn_program_find_type(prog, "struct pid", NULL, &ret->pid_type);
+	if (err)
+		return err;
+	err = drgn_program_find_type(prog, "struct upid", NULL,
+				     &ret->upid_type);
+	if (err)
+		return err;
+	struct drgn_object pidhash_shift;
+	drgn_object_init(&pidhash_shift, prog);
+	err = drgn_program_find_object(prog, "pidhash_shift", NULL,
+				       DRGN_FIND_OBJECT_ANY, &pidhash_shift);
+	if (err)
+		return err;
+	err = drgn_object_read_unsigned(&pidhash_shift, &ret->hash_length);
+	if (err)
+		return err;
+	ret->hash_length = 1 << ret->hash_length;
+	struct drgn_object ns_level;
+	drgn_object_init(&ns_level, prog);
+	err = drgn_object_member_dereference(&ns_level, ns, "level");
+	if (err)
+		return err;
+	uint64_t ns_level_value;
+	err = drgn_object_read_unsigned(&ns_level, &ns_level_value);
+	if (err)
+		return err;
+	snprintf(ret->member_specifier, sizeof(ret->member_specifier),
+		 "numbers[%lu]", ns_level_value);
```

Comment:
Use `PRIu64` for formatting `uint64_t`.

```suggestion
		 "numbers[%" PRIu64 "]", ns_level_value);
```